Framework/Algorithms/src/CalcCountRate.cpp

```diff
@@ -0,0 +1,653 @@
+      }
+
+      // Report progress
+      prog.report(name());
+      PARALLEL_END_INTERUPT_REGION
+    }
+    PARALLEL_CHECK_INTERUPT_REGION
+// calculate final sums
+#pragma omp for
+    for (int64_t j = 0; j < m_numLogSteps; j++) {
+
+      for (int i = 0; i < nThreads; i++) {
+        countRate[j] += Buff[i][j];
+      }
+      // normalize if requested
+      if (!countNormalization.empty()) {
+        countRate[j] /= countNormalization[j];
+      }
+    }
+    if (!countNormalization.empty() && this->buildVisWS()) {
+#pragma omp for
+      for (int64_t j = 0; j < int64_t(m_visNorm.size()); j++) {
+        this->normalizeVisWs(j);
+      }
+    }
+  }
+  // recover histogram timing, but start assuming run starts at
+  std::vector<Kernel::DateAndTime> times(m_numLogSteps);
+
+  double dt = (dTRangeMax - dTRangeMin) / static_cast<double>(m_numLogSteps);
+  auto t0 = m_TRangeMin.totalNanoseconds();
+  for (auto i = 0; i < m_numLogSteps; i++) {
+    times[i] =
+        Kernel::DateAndTime(t0 + static_cast<int64_t>((0.5 + double(i)) * dt));
+  }
+  // store calculated values within the target log.
+  targLog->replaceValues(times, countRate);
+}
+/** hisogram event list into visualization workspace
+* @param el       :: event list to rebin into visualization workspace
+* @param spectraLocks :: pointer to the array of mutexes to lock modifyed
+*                        visualization workspace spectra for a thread
+*/
+void CalcCountRate::hisogramEvents(const DataObjects::EventList &el,
+                                   std::mutex *spectraLocks) {
+
+  if (el.empty())
+    return;
+
+  auto events = el.getEvents();
+  for (const DataObjects::TofEvent &ev : events) {
+    double pulsetime = static_cast<double>(ev.pulseTime().totalNanoseconds());
+    double tof = ev.tof();
+    if (pulsetime < m_visT0 || pulsetime >= m_visTmax)
+      continue;
+    if (tof < m_XRangeMin || tof >= m_XRangeMax)
+      continue;
+
+    size_t n_spec = static_cast<size_t>((pulsetime - m_visT0) / m_visDT);
+    size_t n_bin = static_cast<size_t>((tof - m_XRangeMin) / m_visDX);
+    (spectraLocks + n_spec)->lock();
+    auto &Y = m_visWs->mutableY(n_spec);
+    Y[n_bin]++;
+    (spectraLocks + n_spec)->unlock();
+  }
+}
+
+/** Normalize single spectra of the normalization workspace using prepared
+ * normzlization log
+ @param wsIndex -- appropriate visualization workspace index to normalize
+ spectra
+ */
+void CalcCountRate::normalizeVisWs(int64_t wsIndex) {
+
+  auto Y = m_visWs->dataY(wsIndex);
+  for (auto &yv : Y) {
+    yv /= m_visNorm[wsIndex];
+  }
+}
+/*Analyse input log parameters and logs, attached to the workspace and identify
+ * the parameters of the target log, including experiment time.
+ *
+ @param InputWorkspace -- input workspace to analyse logs
+ */
+void CalcCountRate::setOutLogParameters(
+    const DataObjects::EventWorkspace_sptr &InputWorkspace) {
+
+  std::string NormLogName = getProperty("NormalizationLogName");
```

Comment:
its normal mode of operation to overwrite countingRate log. But indeed, if one tries to overwrite proton_charge and use it for normalization, the result will be bad. Blocking this.